src/include/btree.i

```diff
@@ -1693,3 +1618,154 @@ __wt_split_descent_race(
 	return (pindex != saved_pindex);
 }
+
+/*
+ * __wt_split_prev_race --
+ *	Return if we raced with an internal page split when moving backwards
+ * through the tree.
+ */
+static inline bool
+__wt_split_prev_race(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+	WT_PAGE_INDEX *pindex;
+
+	/*
+	 * There's a split race when a cursor moving backwards through the tree
+	 * descends the tree. If we're splitting an internal page into its
+	 * parent, we move the WT_REF structures and update the parent's page
+	 * index before updating the split page's page index, and it's not an
+	 * atomic update. A thread can read the parent and split page's original
+	 * indexes during a split, or read the parent page's replacement page
+	 * index and then read the split page's original index, either of which
+	 * can lead to skipping pages.
+	 *
+	 * For example, imagine an internal page with 3 child pages, with the
+	 * namespaces a-f, g-h and i-j; the first child page splits. The parent
+	 * starts out with the following page-index:
+	 *
+	 *	| ... | a | g | i | ... |
+	 *
+	 * The split page starts out with the following page-index:
+	 *
+	 *	| a | b | c | d | e | f |
+	 *
+	 * The first step is to move the c-f ranges into a new subtree, so, for
+	 * example we might have two new internal pages 'c' and 'e', where the
+	 * new 'c' page references the c-d namespace and the new 'e' page
+	 * references the e-f namespace. The top of the subtree references the
+	 * parent page, but until the parent's page index is updated, threads in
+	 * the subtree won't be able to ascend out of the subtree. However, once
+	 * the parent page's page index is updated to this:
+	 *
+	 *	| ... | a | c | e | g | i | ... |
+	 *
+	 * threads in the subtree can ascend into the parent. Imagine a cursor
+	 * in the c-d part of the namespace that ascends to the parent's 'c'
+	 * slot. It would then decrement to the slot before the 'c' slot, the
+	 * 'a' slot.
+	 *
+	 * The previous-cursor movement selects the last slot in the 'a' page;
+	 * if the split page's page-index hasn't been updated yet, it selects
+	 * the 'f' slot, which is incorrect. Once the split page's page index is
+	 * updated to this:
+	 *
+	 *	| a | b |
+	 *
+	 * the previous-cursor movement will select the 'b' slot, which is
+	 * correct.
+	 *
+	 * This function takes an argument which is the internal page into which
+	 * we're descending. If the last slot on the page no longer points to
+	 * the current page as its "home", the page is being split and part of
+	 * its namespace moved, we have to restart.
+	 */
+	WT_INTL_INDEX_GET(session, ref->page, pindex);
+	return (pindex->index[pindex->entries - 1]->home != ref->page);
+}
+
+/*
+ * __wt_page_swap_func --
+ *	Swap one page's hazard pointer for another one when hazard pointer
+ * coupling up/down the tree.
+ */
+static inline int
+__wt_page_swap_func(WT_SESSION_IMPL *session,
+    WT_REF *held, WT_REF *want, bool prev_race, uint32_t flags
```

Comment:
Sorry to give the feedback piecemeal, but could we use `LF_ISSET(WT_READ_PREV)` instead of adding a new parameter?

Reply:
@agorrod 
> could we use `LF_ISSET(WT_READ_PREV)` instead of adding a new parameter?

It's not just a cosmetic change.

The "additional parameter version" of the change doesn't check for the race when returning internal pages in a tree ascent. (If we are returning internal pages during a walk (that is, when doing eviction), we hazard-pointer couple to the internal parent of the current page in order to return it, and the changes means we'll detect and restart when the internal parent is splitting during that ascent, where we didn't before. I didn't check before, mostly because I was pretty sure the specific race we were chasing couldn't occur in that case. That said, making the change should avoid eviction skipping big blocks of records in some hot trees, which doesn't seem like a bad thing.)

All that said, I like the idea, it's cleaner than what I did. If it passes my testing, I'll go ahead and push it.

Thanks!